pageserver/src/layered_repository/timeline.rs

```diff
@@ -984,6 +980,22 @@ impl LayeredTimeline {
             if let Some(SearchResult { lsn_floor, layer }) = layers.search(key, cont_lsn)? {
                 //info!("CHECKING for {} at {} on historic layer {}", key, cont_lsn, layer.filename().display());
 
+                //
+                // Is this a remote layer? If so, create a Future that the caller can use to
+                // download the missing layer, and return it.
+                //
+                if let Some(remote_layer) = super::storage_layer::downcast_remote_layer(&layer) {
+                    info!("need remote layer {}", remote_layer.filename().display());
+                    let tl = timeline.myself.upgrade().unwrap();
+
+                    let future = tl.download_remote_layer(Arc::clone(&remote_layer));
+
+                    let dynfuture: std::pin::Pin<
+                        Box<dyn std::future::Future<Output = Result<()>> + Send + Sync>,
+                    > = Box::pin(future);
+                    return Err(PageReconstructError::NeedDownload(dynfuture));
```

Comment:
If we return watch or a join handle (It can be even an enum with Watch/JoinHandle variants and async wait method manual impl Future wrapper that forwards poll to the proper enum variant) we can avoid this pin box stuff around this future